trin-core/src/portalnet/overlay.rs

```diff
@@ -175,12 +190,182 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
+                self.propagate_gossip(content_keys_values);
+            }
+            ProtocolId::State => {
+                // TODO: Verify state content data with Oracle
+
+                // Temporally store content key/value pairs to propagate here
+                let mut content_keys_values: Vec<(StateContentKey, ByteList)> = Vec::new();
+
+                // Iterate over accepted content keys/values and try to store them.
+                for (content_key, content_value) in content_keys.into_iter().zip(content_values) {
+                    let history_key = StateContentKey::try_from(content_key);
+
+                    match history_key {
+                        Ok(key) => {
+                            // Store accepted content in DB
+                            self.store_accepted_content(&key, content_value.clone());
+                            content_keys_values.push((key, content_value))
+                        }
+                        Err(err) => {
+                            warn!("Unable to decode state content key: {err}");
+                            continue;
+                        }
+                    }
+                }
+                // Propagate gossip accepted content
+                self.propagate_gossip(content_keys_values);
+            }
+            _ => return Err(anyhow!("Unsupported content key protocol.")),
+        };
+        Ok(())
+    }
+
+    /// Propagate gossip accepted content via OFFER/ACCEPT:
+    fn propagate_gossip(&self, accepted_content: Vec<(impl OverlayContentKey, ByteList)>) {
+        // Get all nodes from overlay routing table
+        let kbuckets = self.kbuckets.read();
+        let all_nodes: Vec<&kbucket::Node<NodeId, Node>> = kbuckets
+            .buckets_iter()
+            .map(|kbucket| {
+                kbucket
+                    .iter()
+                    .collect::<Vec<&kbucket::Node<NodeId, Node>>>()
+            })
+            .flatten()
+            .collect();
+
+        // HashMap to temporally store all interested nodes and the content
+        let mut nodes_and_content: HashMap<String, Vec<RawContentKey>> = HashMap::new();
+
+        // Filter all nodes from overlay routing table where XOR_distance(content_id, nodeId) < node radius
+        for content in accepted_content {
+            let interested_nodes: Vec<Enr> = all_nodes
+                .clone()
+                .into_iter()
+                .filter(|node| {
+                    XorMetric::distance(&content.0.content_id(), &node.key.preimage().raw())
+                        < node.value.data_radius()
+                })
+                .map(|node| node.value.enr())
+                .collect();
+
+            // Continue if no nodes are interested in the content
+            if interested_nodes.is_empty() {
+                debug!("No nodes interested in neighborhood gossip: content_key={} num_nodes_checked={}", hex_encode(content.0.into()), all_nodes.len());
```

Comment:
nit: this line seems long, and I'm not sure why `cargo fmt` did not pick this up.